# Data Preprocessing:


In [1]:
import pandas as pd
import numpy as np
data=pd.read_csv("C:\\Users\\HP\\OneDrive\\Desktop\\datascience\\Recommendation System\\Recommendation System\\anime.csv")
data

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

checking data structure checking no of riws and columns

In [2]:
data.shape

(12294, 7)

checking features type and non null counts


In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


checking null values

In [4]:
data.isna().sum()

anime_id      0
name          0
genre        62
type         25
episodes      0
rating      230
members       0
dtype: int64

so here genre column have 62 null values, type column 25 null values, rating column have 230 null values 

filling null values 

In [5]:
import warnings
warnings.filterwarnings("ignore")
# Fill missing values
data['genre'].fillna('Unknown', inplace=True)
data['type'].fillna('Unknown', inplace=True)
data['rating'].fillna(data['rating'].mean(), inplace=True)





In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12294 non-null  object 
 3   type      12294 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12294 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


# Feature Extraction:

Decide on the features that will be used for computing similarity (e.g., genres, user ratings).

Convert categorical features into numerical representations if necessary.

Normalize numerical features if required.



In [7]:
data

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

In [8]:
data['genre'].unique()

array(['Drama, Romance, School, Supernatural',
       'Action, Adventure, Drama, Fantasy, Magic, Military, Shounen',
       'Action, Comedy, Historical, Parody, Samurai, Sci-Fi, Shounen',
       ..., 'Hentai, Sports', 'Drama, Romance, School, Yuri',
       'Hentai, Slice of Life'], dtype=object)

In [9]:
data['type'].unique()

array(['Movie', 'TV', 'OVA', 'Special', 'Music', 'ONA', 'Unknown'],
      dtype=object)

In [10]:
print(data['genre'].value_counts())
print(data['type'].value_counts())


genre
Hentai                                                  823
Comedy                                                  523
Music                                                   301
Kids                                                    199
Comedy, Slice of Life                                   179
                                                       ... 
Adventure, Drama, Fantasy, Game, Sci-Fi                   1
Adventure, Demons, Fantasy, Historical                    1
Action, Comedy, Drama, Mecha, Music, Sci-Fi, Shounen      1
Action, Comedy, Fantasy, Mecha, Sci-Fi, Shounen           1
Hentai, Slice of Life                                     1
Name: count, Length: 3265, dtype: int64
type
TV         3787
OVA        3311
Movie      2348
Special    1676
ONA         659
Music       488
Unknown      25
Name: count, dtype: int64


In [11]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import MinMaxScaler


# Convert 'episodes' to numeric (handle 'object' type)
data['episodes'] = pd.to_numeric(data['episodes'], errors='coerce')
data['episodes'].fillna(data['episodes'].mean(), inplace=True) 

# Normalize 'rating' and 'members' columns
scaler = MinMaxScaler()
data[['rating', 'members']] = scaler.fit_transform(data[['rating', 'members']])

# Use TF-IDF for genres (convert to a numerical representation)
tfidf = TfidfVectorizer(stop_words='english')
genre_matrix = tfidf.fit_transform(data['genre'])

# Encode the 'type' column (e.g., 'TV', 'Movie', 'OVA')
label_encoder = LabelEncoder()
data['type_encoded'] = label_encoder.fit_transform(data['type'])

# Combine features into a single dataframe for similarity calculation
features = pd.concat([
    pd.DataFrame(genre_matrix.toarray(), columns=tfidf.get_feature_names_out()),
    data[['rating', 'members', 'type_encoded', 'episodes']]
], axis=1)
features


action  adventure   ai  arts  cars    comedy  dementia  demons  \
0      0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   
1      0.294649   0.317607  0.0   0.0   0.0  0.000000       0.0     0.0   
2      0.250631   0.000000  0.0   0.0   0.0  0.200766       0.0     0.0   
3      0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   
4      0.250631   0.000000  0.0   0.0   0.0  0.200766       0.0     0.0   
...         ...        ...  ...   ...   ...       ...       ...     ...   
12289  0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   
12290  0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   
12291  0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   
12292  0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   
12293  0.000000   0.000000  0.0   0.0   0.0  0.000000       0.0     0.0   

          drama  ecchi  ...  supernatural  thriller  unknown  vampire  yaoi  \
0      0.440247    0.0  ...      0.544416  0.000000      0.0      0.0   0.0   
1      0.335834    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
2      0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
3      0.000000    0.0  ...      0.000000  0.833769      0.0      0.0   0.0   
4      0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
...         ...    ...  ...           ...       ...      ...      ...   ...   
12289  0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
12290  0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
12291  0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
12292  0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   
12293  0.000000    0.0  ...      0.000000  0.000000      0.0      0.0   0.0   

       yuri    rating   members  type_encoded  episodes  
0       0.0  0.924370  0.197872             0       1.0  
1       0.0  0.911164  0.782770             5      64.0  
2       0.0  0.909964  0.112689             5      51.0  
3       0.0  0.900360  0.664325             5      24.0  
4       0.0  0.899160  0.149186             5      51.0  
...     ...       ...       ...           ...       ...  
12289   0.0  0.297719  0.000203             3       1.0  
12290   0.0  0.313325  0.000176             3       1.0  
12291   0.0  0.385354  0.000211             3       4.0  
12292   0.0  0.397359  0.000168             3       1.0  
12293   0.0  0.454982  0.000135             0       1.0  

[12294 rows x 51 columns]

# Recommendation System:

Design a function to recommend anime based on cosine similarity.

Given a target anime, recommend a list of similar anime based on cosine similarity scores.

Experiment with different threshold values for similarity scores to adjust the recommendation list size.


In [12]:
# Check for missing values
features.isnull().sum()

action           0
adventure        0
ai               0
arts             0
cars             0
comedy           0
dementia         0
demons           0
drama            0
ecchi            0
fantasy          0
fi               0
game             0
harem            0
hentai           0
historical       0
horror           0
josei            0
kids             0
life             0
magic            0
martial          0
mecha            0
military         0
music            0
mystery          0
parody           0
police           0
power            0
psychological    0
romance          0
samurai          0
school           0
sci              0
seinen           0
shoujo           0
shounen          0
slice            0
space            0
sports           0
super            0
supernatural     0
thriller         0
unknown          0
vampire          0
yaoi             0
yuri             0
rating           0
members          0
type_encoded     0
episodes         0
dtype: int64

In [13]:
from sklearn.metrics.pairwise import cosine_similarity

# Compute cosine similarity matrix
cos_sim = cosine_similarity(features)

def recommend_anime(target_anime_id, top_n=10):
    # Get index of the target anime
    target_idx = data[data['anime_id'] == target_anime_id].index[0]
    
    
    similarity_scores = cos_sim[target_idx]
    
    # Get indices of the top N similar animes (excluding the target anime itself)
    similar_anime_idx = similarity_scores.argsort()[-top_n-1:-1][::-1]
    
   
    recommended_animes = data.iloc[similar_anime_idx][['name', 'anime_id']]
    recommended_animes['similarity_score'] = similarity_scores[similar_anime_idx]
    
    return recommended_animes

# Example usage: recommend top 5 animes similar to 'Kimi no Na wa.'
recommend_anime(32281, top_n=5)



name  anime_id  similarity_score
1111  Aura: Maryuuin Kouga Saigo no Tatakai     14669          0.972428
1494                               Harmonie     20903          0.940418
208           Kokoro ga Sakebitagatterunda.     28725          0.936365
1959                              Air Movie       713          0.933795
2103                          Clannad Movie      1723          0.896942

# Evaluation:

Split the dataset into training and testing sets.


In [17]:
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(data, test_size=0.2, random_state=42)



Evaluate the recommendation system using appropriate metrics such as precision, recall, and F1-score.

In [18]:
from sklearn.metrics import precision_score, recall_score, f1_score

def evaluate_recommendations(target_anime_id, top_n=10):
    # Get top N recommendations for the target anime
    recommended_animes = recommend_anime(target_anime_id, top_n)
    
    # Get the list of anime ids recommended
    recommended_anime_ids = recommended_animes['anime_id'].tolist()
    
    
    is_target_in_recommendations = 1 if target_anime_id in recommended_anime_ids else 0
    
    
    ground_truth = 1 if target_anime_id in test_data['anime_id'].values else 0
    
    return is_target_in_recommendations, ground_truth


In [19]:
# Evaluate the system
y_true = []
y_pred = []

# Loop over all test data anime
for target_anime_id in test_data['anime_id']:
    is_target_in_recommendations, ground_truth = evaluate_recommendations(target_anime_id, top_n=10)
    y_true.append(ground_truth)
    y_pred.append(is_target_in_recommendations)

# Calculate precision, recall, and F1-score
precision = precision_score(y_true, y_pred)
recall = recall_score(y_true, y_pred)
f1 = f1_score(y_true, y_pred)

print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")


Precision: 1.0000
Recall: 0.0016
F1-Score: 0.0032


Analyze the performance of the recommendation system and identify areas of improvement.

In [22]:
# Loop to evaluate the system with different top_n values
for top_n in [10, 20, 50]:
    y_true = []
    y_pred = []
    
    # Loop through all test data anime
    for target_anime_id in test_data['anime_id']:
        is_target_in_recommendations, ground_truth = evaluate_recommendations(target_anime_id, top_n)
        y_true.append(ground_truth)
        y_pred.append(is_target_in_recommendations)
    
    # Calculate precision, recall, and F1-score for the current top_n
    precision = precision_score(y_true, y_pred)
    recall = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)

    # Print the results
    print(f"Top-N: {top_n}, Precision: {precision:.4f}, Recall: {recall:.4f}, F1-Score: {f1:.4f}")


Top-N: 10, Precision: 1.0000, Recall: 0.0016, F1-Score: 0.0032
Top-N: 20, Precision: 1.0000, Recall: 0.0016, F1-Score: 0.0032
Top-N: 50, Precision: 1.0000, Recall: 0.0016, F1-Score: 0.0032


In [23]:
# Adjusting top_n to see if recall improves
for top_n in [20, 50, 100]:  # Try larger values of top_n
    y_true = []
    y_pred = []
    
    for target_anime_id in test_data['anime_id']:
        is_target_in_recommendations, ground_truth = evaluate_recommendations(target_anime_id, top_n)
        y_true.append(ground_truth)
        y_pred.append(is_target_in_recommendations)
    
    precision = precision_score(y_true, y_pred)
    recall = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)

    print(f"Top-N: {top_n}, Precision: {precision:.4f}, Recall: {recall:.4f}, F1-Score: {f1:.4f}")


Top-N: 20, Precision: 1.0000, Recall: 0.0016, F1-Score: 0.0032
Top-N: 50, Precision: 1.0000, Recall: 0.0016, F1-Score: 0.0032
Top-N: 100, Precision: 1.0000, Recall: 0.0016, F1-Score: 0.0032


# 1. Difference Between User-Based and Item-Based Collaborative Filtering:

User-Based Collaborative Filtering: Recommends items based on the preferences of similar users. If User A likes items that User B likes, items liked by User B will be recommended to User A.

Item-Based Collaborative Filtering: Recommends items similar to the ones the user has already interacted with. If User A liked "Item X", the system recommends similar items to "Item X" based on other users' preferences.

Key Difference:

User-based focuses on user similarity, while item-based focuses on item similarity.

# 2. What is Collaborative Filtering and How Does It Work?

Collaborative Filtering is a recommendation system technique that suggests items based on the behavior of similar users or items. It works in two main ways:

Memory-Based: Uses user-item interaction data to calculate similarities (e.g., user-based or item-based).

Model-Based: Uses machine learning (e.g., matrix factorization) to generate recommendations based on hidden patterns in the data.

How it works:

Collect user interaction data (ratings, clicks, etc.).

Calculate similarity between users or items.

Recommend items based on these similarities.

Advantages:

Does not require item details or user demographics.

Learns and adapts based on user behavior.

Disadvantages:

Cold start problem (new users/items have no data).

Can be computationally expensive.